<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Откройте-таблицу-и-изучите-общую-информацию-о-данных" data-toc-modified-id="Откройте-таблицу-и-изучите-общую-информацию-о-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Откройте таблицу и изучите общую информацию о данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Удаление-пропусков" data-toc-modified-id="Удаление-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Удаление пропусков</a></span></li><li><span><a href="#Обработка-аномальных-значений" data-toc-modified-id="Обработка-аномальных-значений-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обработка аномальных значений</a></span></li><li><span><a href="#Удаление-пропусков-(продолжение)" data-toc-modified-id="Удаление-пропусков-(продолжение)-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Удаление пропусков (продолжение)</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Исследуйте-данные-и-ответьте-на-вопросы" data-toc-modified-id="Исследуйте-данные-и-ответьте-на-вопросы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследуйте данные и ответьте на вопросы</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#Приведите-возможные-причины-появления-пропусков-в-исходных-данных." data-toc-modified-id="Приведите-возможные-причины-появления-пропусков-в-исходных-данных.-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Приведите возможные причины появления пропусков в исходных данных.</a></span></li><li><span><a href="#Объясните,-почему-заполнить-пропуски-медианным-значением-—-лучшее-решение-для-количественных-переменных." data-toc-modified-id="Объясните,-почему-заполнить-пропуски-медианным-значением-—-лучшее-решение-для-количественных-переменных.-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.</a></span></li></ul></li><li><span><a href="#Общий-вывод." data-toc-modified-id="Общий-вывод.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод.</a></span></li></ul></div>

# Проект по исследованию надежности платежеспособности заемщиков

## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`


In [1]:
import pandas as pd

In [2]:
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [15]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

In [20]:
data['total_income_category']

0        B
1        C
2        C
3        B
4        C
        ..
21520    B
21521    C
21522    C
21523    B
21524    C
Name: total_income_category, Length: 21331, dtype: object

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [21]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [22]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [23]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

In [24]:
data['purpose_category']

0        операции с недвижимостью
1          операции с автомобилем
2        операции с недвижимостью
3           получение образования
4              проведение свадьбы
                   ...           
21520    операции с недвижимостью
21521      операции с автомобилем
21522    операции с недвижимостью
21523      операции с автомобилем
21524      операции с автомобилем
Name: purpose_category, Length: 21331, dtype: object

## Исследуйте данные и ответьте на вопросы

### Есть ли зависимость между количеством детей и возвратом кредита в срок?

**Шаг 1.** Для начала проведения анализа, и сравнения влияния количества детей на вероятность возврата кредита в срок, нужно снова обратиться к уникальным значениям столбца `children` для определения категорий.

In [25]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

**Шаг 2.** Отсюда мы выясняем, что имеется разброс количества детей от одного ребенка, до пяти. Теперь нужно найти для каждой категории количество людей просрочивших выплаты в срок.

In [26]:
children_debt = (data[['children','debt']].groupby(['children']).sum())
children_debt

debt
children      
0         1063
1          444
2          194
3           27
4            4
5            0

**Шаг 3.** Мы получили разброс должников по количеству детей в семье. Предварительно можно сделать вывод, что семейные люди более ответственно отсносятся к вопросу выплат. Сейчас нам нужно найти отношение количества должников каждой категории к общему количеству заемщиков. 

In [27]:
debt_sum = data['debt'].count()
debt_sum 

21331

In [28]:
debt_sum_children = (children_debt / debt_sum) * 100 
debt_sum_children.sort_values(by='debt', ascending=False)

debt
children          
0         4.983358
1         2.081478
2         0.909474
3         0.126576
4         0.018752
5         0.000000

In [29]:
data.groupby('children')['debt'].count()

children
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: debt, dtype: int64

In [30]:
data.groupby('children')['debt'].sum()

children
0    1063
1     444
2     194
3      27
4       4
5       0
Name: debt, dtype: int64

In [31]:
children_debt_new = (data.groupby('children')['debt'].sum() / data.groupby('children')['debt'].count()) * 100
children_debt_new

children
0    7.543822
1    9.234609
2    9.454191
3    8.181818
4    9.756098
5    0.000000
Name: debt, dtype: float64

**Вывод:** По отношениям, которые мы получили из таблицы, можно сделать вывод, что больший процент должников - это семьи с 1, 2, 3 и 4 детьми. Явный провал должников мы видим у бездетных, и тех у кого 5 детей в наличии. Стоит уточнить, что в случае с последними, нулевое значение не значит отсутствие займов в принципе. Конкретно у людей с 4 и 5 детьми, несмотря на процент, количество займов и просрочек минимально. Обе категории людей берут кредиты и стараются выплатить их без задержек. Тем не менее случаев по ним слишком мало для объективной оценки. По итогу мы действительно видим влияние количества детей на задержки в оплате, а конкретно случаи с 1 или 2 детьми в семье. Эти люди берут значительно больше кредитов, и чаще доводят их до просрочки, относительно меньшего количества многодетных, и большего количества бездетных, которые в свою очередь, несмотря на кардинально большее количество кредитов, выплачивают их более успешно.

 ### Есть ли зависимость между семейным положением и возвратом кредита в срок?

**Шаг 1.** Для того чтобы найти зависимость между семейным положением и возвратом кредита в срок, мы проведем похожую последовательность действий, чтобы выяснить какая категория людей больше подвержена просрочке. Для начала мы посмотрим какие категории представлены в принципе. 

In [32]:
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

**Шаг 2.** Соответственно для каждой категории мы отыщем количество случаев просрочки оплаты кредита.data['family_status'].unique()

In [33]:
family_status_debt = (data[['family_status','debt',]].groupby(['family_status']).sum())
family_status_debt

debt
family_status              
Не женат / не замужем   273
в разводе                84
вдовец / вдова           63
гражданский брак        385
женат / замужем         927

**Шаг 3.** Как и в прошлом задании, мы подсчитаем отношение.

In [34]:
debt_sum_family_status = (family_status_debt / debt_sum) * 100 
debt_sum_family_status.sort_values(by='debt', ascending=False)

debt
family_status                  
женат / замужем        4.345788
гражданский брак       1.804885
Не женат / не замужем  1.279827
в разводе              0.393793
вдовец / вдова         0.295345

**Вывод:** Из полученного отношения можно сделать вывод, что семьи закрепившие свой брак официально, берут большее количество кредитов, чем все остальные. Они уверенны в завтрашнем дне и стараются благоустроить свою жизнь. Соответсвенно чем большая совокупность количества кредитов, тем больше количество условных задолженностей. Стоит сразу оговорить момент постройки гипотезы. В данном, конкретном контексте, связь уровня просрочек имеет быть непосредственно с положением людей и их доверию друг к другу. Например гражданский брак - это простое сожительство, которое не дает никаких гарантий. Это конечно же не так, но люди склонны думать об позиции именно так. Уровень просрочек почти в два раза ниже, что дает повод предположить, что такие люди берут либо меньше кредитов, либо имеют способность их погасить во время. Законный брак так же предполагает наличие одного или нескольких детей, что также влияет на платежеспособность человека, но опять же это не происходит буквально зависимо друг от друга. Отсюда следует предположение, согласно которому люди состоящие в гражданском браке не обременены, даже условно имея детей, способны покрывать свои долги более эффективно, чем люди состоящие в законном браке. Первые несут меньше ответсвенности перед другим человеком, чем вторые, и поэтому стараются закрыть условный долг быстрее. Так как возможности каждого человека ограничены, условная пара в гражданском браке будет более осторожно обдумывать о взятие кредита, так как в первую очередь будет оцениваться личные возможность каждого из них. Данное предположение можно соответсвенно спроецировать на остальные категории людей. "Не женат/ не замужем", "в разводе" и "вдовец/вдова" аналогично имеют меньшее количество обязательств и так же органичения на платежеспособность, так как наличие детей у данных категорий никто не отменял - они возможны. У данных категорий самый меньший процент просрочек по оплате, как раз связанных с их текущим положением. Они либо не готовы взять на себя такое обязательство, либо не могут выплатить кредит в срок, из - за побочных явлений - например наличия детей. 

Общее количество заемщиков по категориям семейного статуса:

In [35]:
data.groupby('family_status')['debt'].count()

family_status
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
гражданский брак          4134
женат / замужем          12261
Name: debt, dtype: int64

Общее количество должников по категориям семейного статуса:

In [36]:
data.groupby('family_status')['debt'].sum()

family_status
Не женат / не замужем    273
в разводе                 84
вдовец / вдова            63
гражданский брак         385
женат / замужем          927
Name: debt, dtype: int64

In [37]:
family_status_debt_new = (data.groupby('family_status')['debt'].sum() / data.groupby('family_status')['debt'].count()) * 100
family_status_debt_new

family_status
Не женат / не замужем    9.763948
в разводе                7.064760
вдовец / вдова           6.624606
гражданский брак         9.313014
женат / замужем          7.560558
Name: debt, dtype: float64

**Вывод:** Семейный статус также влияет на общее количество невыплат по сроку. Самый низкие показатели мы имеем у вдов и людей в разводе. Также и у семей в официальном браке. В первых двух случаях можно предположить, что данные категории людей не всегда способны взять кредит и соответственно не всегда могут погасить его полностью в срок. На это нам намекают очень малое количество случаев самих займов относительно остальных категорий. Во втором случае люди в браке показывают положительные цифры, так как абсолютно большая часть займов - это именно данная категория. Можно назвать эту категорию самой надежной из всех представленных. Что касается гражданского брака и одиноких людей, можно предположить, что они в основном расчитывают на свои силы, берут в целом меньше займов, но тем не менее не всегда способны выплатить их в срок.

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

**Шаг 1.** Для начала нужно определить, какие категории находятся в базе и их соответсвие разбросу в суммах. 

In [38]:
data['total_income_category'].unique()

array(['B', 'C', 'D', 'E', 'A'], dtype=object)

**Регламент категорий:**

От 0 до 30000 - **E** 


30001 до 50000 - **D**


50001 до 200000 - **С**


200001 до 1000000 - **B**


1000001 и более - **А**

**Шаг 2.** Находим общее количество должников по категориям:

In [39]:
data.groupby('total_income_category')['debt'].count()

total_income_category
A       25
B     5014
C    15921
D      349
E       22
Name: debt, dtype: int64

Распределение должников по уровню заработка:

In [40]:
data.groupby('total_income_category')['debt'].sum()

total_income_category
A       2
B     354
C    1353
D      21
E       2
Name: debt, dtype: int64

**Шаг 3.** Находим отношение количества категорий должников к общему количеству:

In [41]:
total_income_category_debt_new = (data.groupby('total_income_category')['debt'].sum() / data.groupby('total_income_category')['debt'].count()) * 100
total_income_category_debt_new

total_income_category
A    8.000000
B    7.060231
C    8.498210
D    6.017192
E    9.090909
Name: debt, dtype: float64

**Вывод:** Самая большая концентрация должников это E и С категории - люди с заработком от 30000 и 50000 - 200000 соответсвенно. Стоит заметить, что категория Е относительно остальных, представлена очень малым количеством случаев и рассматривать ее спорно. Среди остальных категорий мы можем наблюдать как уровень заработка влияет, как на количество займов, так и на их выплату.

### Как разные цели кредита влияют на его возврат в срок?

**Шаг 1.** Найдем все категории трат на которые дают кредит.

In [42]:
data['purpose_category'].unique()

array(['операции с недвижимостью', 'операции с автомобилем',
       'получение образования', 'проведение свадьбы'], dtype=object)

**Шаг 2.** Находим общее количество задолженностей по категориям целей кредита.

Количество займов на различные цели:

In [43]:
data.groupby('purpose_category')['debt'].count()

purpose_category
операции с автомобилем       4279
операции с недвижимостью    10751
получение образования        3988
проведение свадьбы           2313
Name: debt, dtype: int64

Количество задолженностей на различные цели:

In [44]:
data.groupby('purpose_category')['debt'].sum()

purpose_category
операции с автомобилем      400
операции с недвижимостью    780
получение образования       369
проведение свадьбы          183
Name: debt, dtype: int64

In [45]:
purpose_category_debt_new = (data.groupby('purpose_category')['debt'].sum() / data.groupby('purpose_category')['debt'].count()) * 100
purpose_category_debt_new

purpose_category
операции с автомобилем      9.347978
операции с недвижимостью    7.255139
получение образования       9.252758
проведение свадьбы          7.911803
Name: debt, dtype: float64

**Вывод:** Цели также существенно влияют на количество займов и задолженностей. Сами цели кредитов определяют их важность и сложность в выплатах. По цифрам мы видим, что больше всего займов уходят на недвижимость. Это очень ресурсоемкая деятельность, и вполне очевидно, что очень много людей могут с ней не справится. Тем не менее это далеко не самая большая долговая яма в отношении других категорий. Больше всего должников числятся за займами связанными с покупкой автомобилей и оплатой обучения. Можно предположить, что данные траты не всегда окупают себя, но также требуют внимательности от заемщика. Проведения свадеб достаточно неоднозначная цель. Она не несет никакой особенной пользы, кроме как знакового события. Отсюда на кредит на свадьбу люди смотрят достаточно спорно. В любом случае общее количество займов меньше, чем у остальных категорий, но при этом достаточно высокий уровень просрочек. В отличии от остальных категорий, займы на свадьбы нельзя "отбить", и это своего рода также является причиной их позиции. 

### Приведите возможные причины появления пропусков в исходных данных.

Пропуски в данных замечались только в двух стобцах - это days_employed и total_income. В первом случае это была техническая ошибка, так как изначальные значения имели отрицательное значение. После исправления этих ошибок, пропуски исчезли полностью. Во втором случае могли быть как и тех. ошибка, так и человеческий фактор. Но лично я склоняюсь к первому варианту, так как отсутствие значения в крайне важной для выдачи кредита колонке, является недопустимой ошибкой. Так как сама работа банка подразумевает проверку платежеспособности клиента.

### Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

Медианное значение более корректно выражает значения по выборке в целом. Среднее арифметическое может исказить общие показатели и в итоге дать неверные данные на выходе.

## Общий вывод.

Было проведено исследование влияния различных факторов на платежеспособность клиентов банка и их способность возвращать долги во время. 

Были выдвинуты несколько гипотез для проверки:

1 - Влияние количества детей на платежеспособность заемщика;

2 - Влияние семейного положения на платежеспособность заемщика;

3 - Влияние уровня дохода на платежеспособность заемщика;

4 - Влияние целей кредитов на платежеспособность заемщика.

**1 - Влияние количества детей на платежеспособность заемщика**

Количество детей, и вообще наличие детей влияет на способность заемщика платить во время. Больший уровень задолженностей имеют лица с одним или двумя детьми. Бездетные же имеют намного меньший процент в сравнении с первыми. 

Количество детей:

0 - 7.543822

1 - 9.234609

2 - 9.454191

Как уже было сказано выше, остальные категории имеют слишком малое количество случаев для объективной оценки.

Можно было бы порекомендовать провести исследование с большим количеством выборки в остальных случаях.

**2 - Влияние семейного положения на платежеспособность заемщика**

Самые плохие показатели имеют категории: гражданский брак (9.31%) и одинокие люди (9.7%). На их фоне люди в браке выглядят как надежные, понимащие чего они хотят люди (7.5%). Они несут большую ответственность и поэтому не могут позволить себе лишние штрафные санкции. Остальные группы: в разводе (7%), вдовцы и вдовы (6.6%) - банально берут меньше займов, и если взяли, стараются выплатить их в срок, так как потеря второй половины резко снижает их возможность выплатить кредит во время. 

В данной ситуации можно порекомендовать более тщательно проверять лиц в гражданском браке и одиночек, так как их позиции нестабильны. 

**3 - Влияние уровня дохода на платежеспособность заемщика**

Самые крупные концентрации должников мы наблюдаем у группы B и C. Группы А и Е имеют совсем незначительное количество случаев для объективной оценки. Группа D самая платежеспособная, но и количество займов в целом, у этой группы намного меньше на фоне В и С.

B - 7.06%

C - 8.49%

D - 6.01%

Для более точных выводов можно порекомендовать провести исследование с большим количеством выборки. В данном конкретном случае, даже группа D смотрится не убедительно. Влияние уровня дохода на платежеспособность хоть и имеет место быть, но пока не очевидна.

**4 - Влияние целей кредитов на платежеспособность заемщика**

1 - Задолженности связанные с автомобилем - 9.34%

2 - Задолженности связанные с недвижимостью - 7.25%

3 - Задолженности связанные с получением образования - 9.25%

4 - Задолженности связанные с проведением свадеб - 7.9%

В данной ситуации, можно было бы порекомендовать банку более тщательно проверять людей берущих займы в 3 и 4 случаях. Данные цели ничем не подкреплены, и в случае непредвиденных причин, заемщик не сможет во время выплатить кредит. 

**Дополнительные рекомендации:**

1 - Более тщательно выгружать данные и проверять их перед передачей в работу.

2 - Несколько пунктов показали, что данных недостаточно для объективной оценки, и это может повлиять на выводы.